cheatsheets/File_Upload_Cheat_Sheet.md

```diff
@@ -0,0 +1,163 @@
+# File Upload Cheat Sheet
+
+## Introduction
+
+File upload is becoming more and more an essential part of any application, where the user is able to upload their photo, their CV, or a video showcasing a project they are working on. It is evident that applications should be able to handle file uploads based on the business requirement. In return, the application should be able to fend off bogus, and malicious files in a way to keep the application and the users safe.
+
+In short, the following principles should be followed to reach a secure file upload implementation:
+
+- **Whitelist allowed extensions. Only allow safe and critical extensions for business functionality**
+  - **Ensure that [input validation](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Input_Validation_Cheat_Sheet.md#validating-free-form-unicode-text) is applied before validating the extensions.**
+- **Allowed users only can upload files**
+- **Validate the file type, don't trust the [Content-Type header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) as it can be spoofed**
+- **Change the file name to something generated by the application**
+- **Set a file name length limit. Restrict the allowed characters if possible**
+  - **Setting a character limit could deny the attacker from sending null bytes, adding secondary extensions, and performing XSS attacks if the file name is reflected**
+- **Set a file size limit**
+- **Store the files on a different server. If that's not possible, store them in a private directory**
+  - **In the case of public access to the files, ensure that proper permissions are set, file names are safe, and encoded**
```

Comment:
Usually better to serve through a file download handler so you can do access control/etc

Reply:
What do you mean by that?

Reply:
So having a `download.php?file=someid` is usually better than just having `/uploads/file.docx`, because you can do access control/validation/etc on the handler. Directly accessible files are much harder to do access control for.